src/xb-opcode.c

```diff
@@ -413,6 +440,17 @@ xb_opcode_set_val (XbOpcode *self, guint32 val)
 	self->val = val;
 }
 
+/* private */
+void
+xb_opcode_set_token (XbOpcode *self, guint idx, const gchar *val)
+{
+	g_return_if_fail (val != NULL);
+	g_return_if_fail (val[0] != '\0');
+	g_return_if_fail (idx < XB_OPCODE_TOKEN_MAX);
+	self->tokens[idx] = val;
```

Comment:
`self->tokens` is treated as a `NULL`-terminated array elsewhere in this file, so this function should ensure the `NULL` terminator is correct. That can’t happen if, for example, a caller calls `set_token(0, "foo")` then `set_token(15, "disjoint")`. It doesn’t happen now, but could happen later on if the code is refactored.

My suggestion would be to rename this `xb_opcode_append_token()` and drop the `idx` argument. Return a `gboolean` which is `TRUE` for success, or `FALSE` if the tokens array is full. That would simplify error handling in the callers (they would no longer have to check `XB_OPCODE_TOKEN_MAX` themselves) and might mean you could bring `XB_OPCODE_TOKEN_MAX` into this C file as a private implementation detail (although I haven’t worked through whether that’s possible).

Reply:
> My suggestion would be to rename this xb_opcode_append_token() and drop the idx argument

Would we have to scan all the elements before adding each one? Doesn't that make each append check the NULLness a factorial number of times? Or do we make _XbOpcode slightly larger and store a `tokens_len` kind of thing?

Reply:
Probably add a `tokens_len`

Reply:
Done!